src/macros-by-example.adoc

```diff
@@ -0,0 +1,820 @@
+
+[{mrk}]
+----
+(*public* temperature
+  [(*decimal* degrees), (*symbol?* scale)]
+  {degrees: degrees, scale: scale})
+----
+
+Since the scale is voidable, we can pass it void:
+
+----
+(:temperature 96 F)    ⇒ {degrees:96, scale:F}
+(:temperature 283 (:)) ⇒ {degrees:283}
+----
+
+Note that the result’s `scale` field has disappeared because no value was provided.  It would be
+more useful to fill in a default value, and to do that we introduce another special form that can
+detect void:
+
+[{mrk}]
+----
+(*public* temperature
+  [(*decimal* degrees), (*symbol?* scale)]
+  {degrees: degrees, scale: (*if_void* scale (*literal* K) scale)})
+----
+----
+(:temperature 96 F)    ⇒ {degrees:96,  scale:F}
+(:temperature 283 (:)) ⇒ {degrees:283, scale:K}
+----
+
+The `*if_void*` form is if/then/else syntax testing stream emptiness. It has three sub-expressions,
+the first being a stream to check. If and only if that stream is void (_i.e._, it produces no
+values), the second sub-expression is expanded and its results are returned by the `*if_void*`
+expression. Otherwise, it produces at least one value, so the third sub-expression is expanded and
+returned.
+
+NOTE: Exactly one branch is expanded, because otherwise the void stream might be used in a context
+that requires a value, resulting in an errant expansion error.
+
+To refine things a bit further, trailing voidable arguments can be omitted entirely:
+
+----
+(:temperature 283) ⇒ {degrees:283, scale:K}
+----
+
+
+==== zero-or-more
+
+A zero-or-more `***` parameter is identical in meaning to our friend `*\...*` from the perspective
+of its template, but it can be used in any position, not just last place.
+
+[{mrk}]
+----
+(*public* prices
+  [(*number** amount), (*symbol* currency)]
+  (*for_each* amount
+    (price amount currency)))
+----
+
+From the perspective of the invoker, the calling convention is different since the “all the rest”
+convention can’t be used to draw the boundaries of the stream.  Instead, we use a list or
+S-expression as a delimiting container to group the applicable sub-expressions:
+
+[{mrk}]
+----
+(:prices () JPY)         ⇒ _void_
+(:prices (10  9.99) GBP) ⇒ {amount:10, currency:GBP} {amount:9.99, currency:GBP}
+(:prices [10, 9.99] GBP) ⇒ {amount:10, currency:GBP} {amount:9.99, currency:GBP}
+----
+
+The delimiting sequence is effectively shorthand for a `values` expression, but arbitrary
+expressions are not allowed:
+
+[{mrk}]
+----
+(:prices (:values 10 9.99 12.) GBP) ⇒ _**error**: delimiting list or sexp expected_
+(:prices (:) GBP)                   ⇒ _**error**: delimiting list or sexp expected_
+----
+
+That’s because the binary representation of these parameters uses a tagless format for these
+delimiters to keep the common case as dense as possible, and it’s not possible to replace that
+container with a macro invocation.  The text form mirrors that limitation, but you can always
+call a macro inside the delimiter, with no loss of generality:
```

Comment:
It is unclear to me why the text syntax needs to have this limitation--it seems arbitrary to me.

Reply:
I'm trying to prevent legal expressions in text that don't have corresponding binary expressions (and vice versa).  The binary form always has the delimiting syntax.

That said, I do want to reconsider these things after I have a better grip on the binary encoding.